File: android-native/app/src/main/java/com/daidai/panel/ui/deps/DepViewModel.kt
package com.daidai.panel.ui.deps

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.daidai.panel.core.network.ApiEndpoints
import com.daidai.panel.core.network.NetworkModule
import com.daidai.panel.data.model.Dependency
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

data class DepListState(
    val deps: List<Dependency> = emptyList(),
    val pythonRuntimes: List<Map<String, Any>> = emptyList(),
    val total: Int = 0,
    val isLoading: Boolean = false,
    val installingDeps: Set<String> = emptySet(),
    val error: String? = null,
    val installLog: String = "",
    val selectedTab: Int = 0
)

@HiltViewModel
class DepViewModel @Inject constructor(
    private val networkModule: NetworkModule
) : ViewModel() {

    private val _state = MutableStateFlow(DepListState())
    val state: StateFlow<DepListState> = _state.asStateFlow()

    init {
        load()
        loadPythonRuntimes()
    }

    fun load() {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, error = null)
            try {
                val api = networkModule.getApiService()
                val params = mutableMapOf<String, String>()
                if (_state.value.selectedTab == 0) params["type"] = "nodejs"
                else params["type"] = "python"
                val response = api.getDependencies(params)
                if (response.isSuccessful && response.body()?.isSuccess == true) {
                    _state.value = _state.value.copy(
                        deps = response.body()?.data ?: emptyList(),
                        total = response.body()?.total ?: 0,
                        isLoading = false
                    )
                } else {
                    _state.value = _state.value.copy(
                        isLoading = false,
                        error = response.body()?.message ?: "加载失败"
                    )
                }
            } catch (e: Exception) {
                _state.value = _state.value.copy(
                    isLoading = false,
                    error = e.message ?: "网络错误"
                )
            }
        }
    }

    fun loadPythonRuntimes() {
        viewModelScope.launch {
            try {
                val api = networkModule.getApiService()
                val response = api.getPythonRuntimes()
                if (response.isSuccessful && response.body()?.isSuccess == true) {
                    @Suppress("UNCHECKED_CAST")
                    _state.value = _state.value.copy(
                        pythonRuntimes = response.body()?.data as? List<Map<String, Any>> ?: emptyList()
                    )
                }
            } catch (_: Exception) {}
        }
    }

    fun selectTab(tab: Int) {
        _state.value = _state.value.copy(selectedTab = tab)
        load()
    }

    fun install(name: String, type: String) {
        viewModelScope.launch {
            val installing = _state.value.installingDeps.toMutableSet()
            installing.add(name)
            _state.value = _state.value.copy(installingDeps = installing, error = null)
            try {
                val api = networkModule.getApiService()
                if (type == "pip") {
                    api.pipInstall(mapOf("name" to name))
                } else {
                    api.npmInstall(mapOf("name" to name))
                }
                streamInstallLog(name)
                load()
            } catch (e: Exception) {
                _state.value = _state.value.copy(error = e.message ?: "安装失败")
            } finally {
                val updated = _state.value.installingDeps.toMutableSet()
                updated.remove(name)
                _state.value = _state.value.copy(installingDeps = updated)
            }
        }
    }

    private fun streamInstallLog(name: String) {
        viewModelScope.launch {
            try {
                val api = networkModule.getApiService()
                val response = api.streamDepLog(ApiEndpoints.depLogStream(0) + "?name=$name")
                if (response.isSuccessful) {
                    val body = response.body()
                    val log = body?.string() ?: ""
                    _state.value = _state.value.copy(installLog = log)
                }
            } catch (_: Exception) {}
        }
    }

    fun uninstall(id: Int) {
        viewModelScope.launch {
            try {
                val api = networkModule.getApiService()
                api.deleteDependency(id)
                load()
            } catch (_: Exception) {}
        }
    }

    fun reinstall(id: Int) {
        viewModelScope.launch {
            try {
                val api = networkModule.getApiService()
                api.reinstallDep(ApiEndpoints.depReinstall(id))
                load()
            } catch (_: Exception) {}
        }
    }

    fun cancel(name: String) {
        viewModelScope.launch {
            try {
                val api = networkModule.getApiService()
                api.cancelDep(ApiEndpoints.depCancel(0) + "?name=$name")
                val updated = _state.value.installingDeps.toMutableSet()
                updated.remove(name)
                _state.value = _state.value.copy(installingDeps = updated)
            } catch (_: Exception) {}
        }
    }

    fun clearError() {
        _state.value = _state.value.copy(error = null)
    }
}

File: ios-native/DaidaiPanel/ui/deps/DepViewModel.swift
import Foundation

@MainActor
final class DepViewModel: ObservableObject {
    @Published var deps: [Dependency] = []
    @Published var isLoading = false
    @Published var error: String?
    @Published var selectedTab: DepTab = .nodejs

    private var api: ApiService

    enum DepTab: String, CaseIterable {
        case nodejs = "Node.js"
        case python = "Python"
    }

    init(api: ApiService) {
        self.api = api
    }

    func updateAPI(_ api: ApiService) {
        self.api = api
    }

    func load() async {
        isLoading = true
        error = nil
        let type = selectedTab == .nodejs ? "nodejs" : "python"
        do {
            let response: ApiResponse<PaginatedData<Dependency>> = try await api.getDeps(page: 1, pageSize: 100, type: type)
            self.deps = response.data?.items ?? []
        } catch {
            self.error = ApiUtils.extractErrorMessage(from: error)
        }
        isLoading = false
    }

    func filteredDeps() -> [Dependency] {
        deps
    }

    func install(name: String, version: String? = nil) async throws {
        let type = selectedTab == .nodejs ? "nodejs" : "python"
        var body: [String: Any] = ["name": name, "type": type]
        if let version, !version.isEmpty { body["version"] = version }
        let _: ApiResponse<Dependency> = try await api.createDep(body)
        await load()
    }

    func uninstall(_ id: Int) async throws {
        let _: ApiResponse<EmptyData> = try await api.deleteDep(id)
        deps.removeAll { $0.id == id }
    }

    func reinstall(_ id: Int) async throws {
        let _: ApiResponse<EmptyData> = try await api.reinstallDep(id)
        await load()
    }

    func cancel(_ id: Int) async throws {
        let _: ApiResponse<EmptyData> = try await api.cancelDep(id)
        await load()
    }
}
